product/crud/handlers/handle_create_product.py

```diff
@@ -1,51 +1,46 @@
-from http import HTTPStatus
-from typing import Any, Dict
+from http import HTTPMethod
+from typing import Any
 
 from aws_lambda_env_modeler import get_environment_variables, init_environment_variables
+from aws_lambda_powertools.logging import correlation_paths
 from aws_lambda_powertools.metrics import MetricUnit
-from aws_lambda_powertools.utilities.parser import ValidationError, parse
 from aws_lambda_powertools.utilities.typing import LambdaContext
 
 from product.crud.domain_logic.create_product import create_product
+from product.crud.handlers.constants import PRODUCT_PATH
 from product.crud.handlers.schemas.env_vars import CreateVars
-from product.crud.handlers.utils.http_responses import build_response
 from product.crud.handlers.utils.observability import logger, metrics, tracer
-from product.crud.schemas.exceptions import InternalServerException, ProductAlreadyExistsException
-from product.crud.schemas.input import CreateProductRequest
+from product.crud.handlers.utils.rest_api_resolver import app
+from product.crud.schemas.input import CreateProductInput
 from product.crud.schemas.output import CreateProductOutput
 
 
-@init_environment_variables(model=CreateVars)
-@metrics.log_metrics
-@tracer.capture_lambda_handler(capture_response=False)
-def handle_create_product(event: Dict[str, Any], context: LambdaContext) -> Dict[str, Any]:
-    logger.set_correlation_id(context.aws_request_id)
-
+@app.route(PRODUCT_PATH, method=HTTPMethod.PUT)
+def handle_create_product(product_id: str) -> dict[str, Any]:
     env_vars: CreateVars = get_environment_variables(model=CreateVars)
     logger.debug('environment variables', extra=env_vars.model_dump())
 
-    try:
-        # we want to extract and parse the HTTP body from the api gw envelope
-        create_input: CreateProductRequest = parse(event=event, model=CreateProductRequest)
-        logger.info('got create product request', extra={'product': create_input.model_dump()})
-    except (ValidationError, TypeError) as exc:  # pragma: no cover
-        logger.exception('event failed input validation', extra={'error': str(exc)})
-        return build_response(http_status=HTTPStatus.BAD_REQUEST, body={})
+    # we want to extract and parse the HTTP body from the api gw envelope
+    payload = app.current_event.body if app.current_event.body is not None else ''
+    create_input: CreateProductInput = CreateProductInput.model_validate_json(payload)
```

Comment:
Is mypy unable to infer that `create_input` will have the type of `CreateProductInput`?

*for my own learning with Pydantic V2, despite having Pydantic plugin for mypy now.

Reply:
it's not about that, it's because model_validate_json expects a str or bytes but the payload can be None. so i added the ugly if so there's always a string value